docs/sassc.1
.TH SASSC 1
.SH NAME
sassc \- HTML5 based management tool for KVM                                        
.SH SYNOPSIS                                                                         
.B sassc
[\fB--version\fP] [\fB-h\fP|\fB--help\fP] [\fB-s\fP|\fB--output-style=\fP
\fIcoding_style\fP] [\fB-m\fP|\fB-g\fP|\fB--sourcemap\fP] [\fB-I\fP|
\fB--include-path=\fP \fIdirectory\fP] [\fB-w\fP|\fB--watch\fP] [\fB-p\fP|
\fB--precision=\fP \fInumber\fP] \fISCSS_FILE\fP [\fIOUT_CSS_FILE\fP]
.SH DESCRIPTION                                                                      
\fBsassc\fP provides a SassC compliant command line interface (SassC: 
https://github.com/sass/sassc).
.br
It is based on the python libsass (https://github.com/dahlia/libsass-python)
library. See also libsass : http://github.com/sass/libsass.
.br
Sass is a CSS pre-processor language to add on exciting, new, awesome features
to CSS. Sass was the first language of its kind and by far the most mature and
up to date codebase.
.br
For more information about Sass itself, please visit http://sass-lang.com
.SH OPTIONS
The following options are supported:
.TP
\fB--version\fP
Show program's version number and exit.
.TP
\fB-h\fP, \fB--help\fP
Show this help message and exit.
.TP
\fB-s\fP, \fB--output-style=\fP \fIcoding_style\fP
Coding style of the compiled result.  Choose one of compact, expanded,
compressed, or nested (default: \fInested\fP).
.TP
\fB-m\fP, \fB-g\fP|\fB--sourcemap\fP
Emit source map.  Requires the second argument (output css filename).
.TP
\fB-I\fP, \fB--include-path=\fP \fIdirectory\fP
Path to find "@import"ed (S)CSS source files.  Can be multiply used.
.TP
\fB-w\fP, \fB--watch\fP
Watch file for changes.  Requires the second argument (output css filename).
.TP
\fB-p\fP, \fB--precision=\fP \fInumber\fP
Set the precision for numbers (default: \fI5\fP).

.SH LICENCE
.br
sassc and python-libsass is provided under MIT license.
.br
Copyright (c) 2015 Hong Minhee <http://hongminhee.org/>
.SH BUGS
Current bugs can be found here : https://github.com/dahlia/libsass-python/issues
.br
If you find any, please open an issue :
https://github.com/dahlia/libsass-python/issues/new
.SH AUTHOR
\fBHong Minhee\fP <hongminhee@member.fsf.org>


docs/changes.rst
Changelog
=========

Version 0.9.2
-------------

Released on November 12, 2015.

- Follow up the libsass upstream: 3.3.2 --- See the release notes of Libsass
  3.3.2__. [by Anthony Sottile]
- Require VS 2015 to build on windows [:issues:`99` by Anthony Sottile]

__ https://github.com/sass/libsass/releases/tag/3.3.2

Version 0.9.1
-------------

Released on October 29, 2015.

- Follow up the libsass upstream: 3.3.1 --- See the release notes of Libsass
  3.3.1__. [by Anthony Sottile]

__ https://github.com/sass/libsass/releases/tag/3.3.1


Version 0.9.0
-------------

Released on October 28, 2015.

- Fix a bug with writing UTF-8 to a file [:issue:`72` by Caleb Ely]
- Fix a segmentation fault on ^C [:issues:`87` by Anthony Sottile]
- Follow up the libsass upstream: 3.3.0 --- See the release notes of Libsass
  3.3.0__. [:issues:`96` by Anthony Sottile]

__ https://github.com/sass/libsass/releases/tag/3.3.0


Version 0.8.3
-------------

Released on August 2, 2015.

- Follow up the libsass upstream: 3.2.5 --- See the release notes of Libsass
  3.2.5__.  [:issue:`79`, :issue:`80` by Anthony Sottile]
- Fixed a bug that :file:`*.sass` files were ignored.
  [:issue:`78` by Guilhem MAS-PAITRAULT]

__ https://github.com/sass/libsass/releases/tag/3.2.5


Version 0.8.2
-------------

Released on May 19, 2015.

- Follow up the libsass upstream: 3.2.4 --- See the release notes of Libsass
  3.2.3__, and 3.2.4__.  [:issue:`69` by Anthony Sottile]
- The default value of :class:`~sassutils.wsgi.SassMiddleware`'s
  ``error_status`` parameter was changed from ``'500 Internal Server Error'``
  to ``'200 OK'`` so that Mozilla Firefox can render the error message well.
  [:issue:`67`, :issue:`68`, :issue:`70` by zxv]

__ https://github.com/sass/libsass/releases/tag/3.2.3
__ https://github.com/sass/libsass/releases/tag/3.2.4


Version 0.8.1
-------------

Released on May 14, 2015.

- Fixed a bug that there was no ``'expanded'`` in :const:`sass.OUTPUT_STYLES`
  but ``'expected'`` instead which is a typo.  [:issue:`66` by Triangle717]
- Fixed broken FreeBSD build.  [:issue:`65` by Toshiharu Moriyama]


Version 0.8.0
-------------

Released on May 3, 2015.

- Follow up the libsass upstream: 3.2.2 --- See the release notes of Libsass
  3.2.0__, 3.2.1__, and 3.2.2__.
  [:issue:`61`, :issue:`52`, :issue:`56`, :issue:`58`, :issue:`62`, :issue:`64`
  by Anthony Sottile]

  - Compact and expanded output styles  [:issue:`37`]
  - Strings and interpolation closer to Ruby Sass
  - The correctness of the generated sourcemap files
  - Directive buddling
  - Full support for the ``@at-root`` directive
  - Full support for ``!global`` variable scoping

- Now underscored files are ignored when compiling a directory.
  [:issue:`57` by Anthony Sottile]
- Fixed broken FreeBSD build.  [:issue:`34`, :issue:`60` by Ilya Baryshev]
- :class:`~sassutils.wsgi.SassMiddleware` became to log syntax errors
  if exist during compilation to ``sassutils.wsgi.SassMiddleware`` logger
  with level ``ERROR``.  [:issue:`42`]

__ https://github.com/sass/libsass/releases/tag/3.2.0
__ https://github.com/sass/libsass/releases/tag/3.2.1
__ https://github.com/sass/libsass/releases/tag/3.2.2


Version 0.7.0
-------------

Released on March 6, 2015.

Anthony Sottile contributed to the most of this release.  Huge thanks to him!

- Follow up the libsass upstream: 3.1.0 --- See the `release note`__ of Libsass.
  [:issue:`38`, :issue:`43` by Anthony Sottile]

  - Custom functions and imports
  - Decrementing in ``@for`` loops
  - ``@debug`` and ``@error``
  - ``not`` operator
  - ``nth()`` for maps
  - ``inspect()``
  - ``feature-exists()``
  - ``unique-id()``
  - ``random()``

- Added custom functions support.  [:issue:`13`, :issue:`44` by Anthony Sottile]

  - Added :class:`sass.SassFunction` class.
  - Added ``custom_functions`` parameter to :func:`sass.compile()` function.
  - Added data types for custom functions:

    - :class:`sass.SassNumber`
    - :class:`sass.SassColor`
    - :class:`sass.SassList`
    - :class:`sass.SassMap`
    - :class:`sass.SassError`
    - :class:`sass.SassWarning`

- Added ``precision`` parameter to :func:`sass.compile()` function.
  [:issue:`39` by Andrea Stagi]
- :program:`sassc` has a new :option:`-p <sassc -p>`/:option:`--precision
  <sassc --precision>` option.  [:issue:`39` by Andrea Stagi]

__ https://github.com/sass/libsass/releases/tag/3.1.0


Version 0.6.2
-------------

Released on November 25, 2014.

Although 0.6.0--0.6.1 have needed GCC (G++) 4.8+, LLVM Clang 3.3+,
now it became back to only need GCC (G++) 4.6+, LLVM Clang 2.9+,
or Visual Studio 2013 Update 4+.

- Follow up the libsass upstream: 3.0.2 --- See the `release note`__ of libsass.
  [:issue:`33` by Rodolphe Pelloux-Prayer]
- Fixed a bug that :program:`sassc --watch` crashed when a file is not
  compilable on the first try.  [:issue:`32` by Alan Justino da Silva]
- Fixed broken build on Windows.

__ https://github.com/sass/libsass/releases/tag/3.0.2


Version 0.6.1
-------------

Released on November 6, 2014.

- Follow up the libsass upstream: 3.0.1 --- See the `release note`__ of Libsass.
- Fixed a bug that :class:`~sassutils.wsgi.SassMiddleware` never closes
  the socket on some WSGI servers e.g. ``eventlet.wsgi``.

__ https://github.com/sass/libsass/releases/tag/3.0.1


Version 0.6.0
-------------

Released on October 27, 2014.

Note that since libsass-python 0.6.0 (and libsass 3.0) it requires C++11
to compile.  Although 0.6.2 became back to only need GCC (G++) 4.6+,
LLVM Clang 2.9+, from 0.6.0 to 0.6.1 you need GCC (G++) 4.8+, LLVM Clang 3.3+,
or Visual Studio 2013 Update 4+.

- Follow up the libsass upstream: 3.0 --- See the `release note`__ of Libsass.

  - Decent extends support
  - Basic Sass Maps Support 
  - Better UTF-8 Support
  - ``call()`` function
  - Better Windows Support
  - Spec Enhancements

- Added missing `partial import`_ support.  [:issue:`27` by item4]
- :const:`~sass.SOURCE_COMMENTS` became deprecated.

sassc.py
#!/usr/bin/env python
""":mod:`sassc` --- SassC compliant command line interface
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This provides SassC_ compliant CLI executable named :program:`sassc`:

.. sourcecode:: console

   $ sassc
   Usage: sassc [options] SCSS_FILE [CSS_FILE]

There are options as well:

.. option:: -s <style>, --output-style <style>

   Coding style of the compiled result.  The same as :func:`sass.compile()`
   function's ``output_style`` keyword argument.  Default is ``nested``.

.. option:: -I <dir>, --include-path <dir>

   Optional directory path to find ``@import``\ ed (S)CSS files.
   Can be multiply used.

.. option:: -m, -g, --sourcemap

   Emit source map.  Requires the second argument (output CSS filename).
   The filename of source map will be the output CSS filename followed by
   :file:`.map`.

   .. versionadded:: 0.4.0

.. option:: -w, --watch

   Watch file for changes.  Requires the second argument (output CSS
   filename).

   .. versionadded:: 0.4.0

.. option:: -p, --precision

   Set the precision for numbers. Default is 5.

   .. versionadded:: 0.7.0

.. option:: -v, --version

   Prints the program version.

.. option:: -h, --help

   Prints the help message.

.. _SassC: https://github.com/sass/sassc

"""
from __future__ import print_function

import functools
import io
import optparse
import os
import sys
import time

from sass import __version__ as VERSION, OUTPUT_STYLES, CompileError, compile


def main(argv=sys.argv, stdout=sys.stdout, stderr=sys.stderr):
    parser = optparse.OptionParser(
        usage='%prog [options] SCSS_FILE [OUT_CSS_FILE]',
        version='%prog ' + VERSION
    )
    output_styles = list(OUTPUT_STYLES)
    output_styles = ', '.join(output_styles[:-1]) + ', or ' + output_styles[-1]
    parser.add_option('-s', '--output-style', metavar='STYLE', type='choice',
                      choices=list(OUTPUT_STYLES), default='nested',
                      help='Coding style of the compiled result.  Choose one '
                           'of ' + output_styles + '. [default: %default]')
    parser.add_option('-m', '-g', '--sourcemap', dest='source_map',
                      action='store_true', default=False,
                      help='Emit source map.  Requires the second argument '
                           '(output css filename).')
    parser.add_option('-I', '--include-path', metavar='DIR',
                      dest='include_paths', action='append',
                      help='Path to find "@import"ed (S)CSS source files.  '
                           'Can be multiply used.')
    parser.add_option('-w', '--watch', action='store_true',
                      help='Watch file for changes.  Requires the second '
                           'argument (output css filename).')
    parser.add_option(
        '-p', '--precision', action='store', type='int', default=5,
        help='Set the precision for numbers. [default: %default]'
    )
    options, args = parser.parse_args(argv[1:])
    error = functools.partial(print,
                              parser.get_prog_name() + ': error:',
                              file=stderr)
    if not args:
        parser.print_usage(stderr)
        error('too few arguments')
        return 2
    elif len(args) > 2:
        parser.print_usage(stderr)
        error('too many arguments')
        return 2
    filename = args[0]
    if options.source_map and len(args) < 2:
        parser.print_usage(stderr)
        error('-m/-g/--sourcemap requires the second argument, the output '
              'css filename.')
        return 2
    elif options.watch and len(args) < 2:
        parser.print_usage(stderr)
        error('-w/--watch requires the second argument, the output css '
              'filename.')
        return 2
    else:
        pass
    while True:
        try:
            mtime = os.stat(filename).st_mtime
            if options.source_map:
                source_map_filename = args[1] + '.map'  # FIXME
                css, source_map = compile(
                    filename=filename,
                    output_style=options.output_style,
                    source_map_filename=source_map_filename,
                    include_paths=options.include_paths,
                    precision=options.precision
                )
            else:
                source_map_filename = None
                source_map = None
                css = compile(
                    filename=filename,
                    output_style=options.output_style,
                    include_paths=options.include_paths,
                    precision=options.precision
                )
        except (IOError, OSError) as e:
            error(e)
            return 3
        except CompileError as e:
            error(e)
            if not options.watch:
                return 1
        except KeyboardInterrupt:
            break
        else:
            if len(args) < 2:
                print(css, file=stdout)
            else:
                with io.open(args[1], 'w', encoding='utf-8') as f:
                    f.write(css)
                if options.watch:
                    print(filename, 'is just compiled to', args[1],
                          file=stdout)
            if source_map_filename:
                with io.open(source_map_filename, 'w', encoding='utf-8') as f:
                    f.write(source_map)
        if options.watch:  # pragma: no cover
            # FIXME: we should utilize inotify on Linux, and FSEvents on Mac
            while True:
                try:
                    st = os.stat(filename)
                    if st.st_mtime > mtime:
                        print(filename, 'changed; recompile...', file=stdout)
                        break
                    time.sleep(0.5)
                except KeyboardInterrupt:
                    return 0
        else:
            break
    return 0


if __name__ == '__main__':
    sys.exit(main())
